### PR TITLE
Adding support for docker volume driver

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Support docker volume driver in task definition. This PR is dependent on #1406, please review #1406 first.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 
yes
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/taskvolume_test.go [hunk 1]
```diff
@@ -19,52 +19,305 @@ import (
...
+	assert.True(t, ok, "Expected v2 to be host volume")
+	assert.Equal(t, "/path", v2.Volume.(*taskresourcevolume.FSHostVolume).FSSourcePath, "Unmarshaled v2 didn't match marshalled v2")
+
+	dockerVolume, ok := v3.Volume.(*taskresourcevolume.DockerVolumeConfig)
+	assert.True(t, ok, "incorrect DockerVolumeConfig type")
+	assert.Equal(t, "task", dockerVolume.Scope)
+	assert.Equal(t, "local", dockerVolume.Driver)
+}
+
+func TestInitializeDockerLocalDockerVolume(t *testing.T) {
```
Comment (yunhee-l) [inline agent/api/task/taskvolume_test.go]: nit: DockerLocalDocker -> LocalDocker

File: agent/api/task/task.go [hunk 2]
```diff
@@ -261,6 +267,118 @@ func (task *Task) volumeName(name string) string {
...
+	}
+
+	vol.Volume = &volumeResource.VolumeConfig
+	task.AddResource(resourcetype.DockerVolumeKey, volumeResource)
+	task.updateContainerVolumeDependency(vol.Name)
+	return nil
+}
+
+// addSharedVolumes adds shared volume into task resources and updates container dependency
+func (task *Task) addSharedVolumes(ctx context.Context, dockerClient dockerapi.DockerClient,
```
Comment (yunhee-l) [inline agent/api/task/task.go]: how is shared volume's cleanup handled? we need to treat this differently from task scoped volume whose resourceStatusToTransition specifies cleanup at container cleanup.

Comment (richardpen) [reply agent/api/task/task.go]: yes, the shared volume won't be cleaned up [here](https://github.com/aws/amazon-ecs-agent/blob/29015316435f0b640c3477e1a993f88302e1092a/agent/taskresource/volume/dockervolume.go#L284-L289)

File: agent/api/task/task.go [hunk 3]
```diff
@@ -261,6 +267,118 @@ func (task *Task) volumeName(name string) string {
...
+	vol *TaskVolume) error {
+
+	volumeConfig := vol.Volume.(*taskresourcevolume.DockerVolumeConfig)
+	volumeConfig.DockerVolumeName = vol.Name
+	if volumeConfig.Autoprovision {
+		volumeMetadata := dockerClient.InspectVolume(ctx, vol.Name, dockerapi.InspectContainerTimeout)
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/task/task.go]: Missed that.

File: agent/app/agent_capability.go [hunk 4]
```diff
@@ -210,6 +213,7 @@ func (agent *ecsAgent) appendVolumeDriverCapabilities(capabilities []*ecs.Attrib
 		seelog.Warnf("Scanning plugins failed: %v", err)
 		// do not return yet, we need the list of plugins below. range handles nil slice.
 	}
+	capabilities = appendNameOnlyAttribute(capabilities, attributePrefix+capabilityDockerVolumeDriverInfix+volume.DockerLocalVolumeDriver)
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/app/agent_capability.go]: I think we have communicated to the customer the old docker version should be deprecated, but I will check with the team.

File: agent/api/task/task.go [hunk 5]
```diff
@@ -261,6 +267,120 @@ func (task *Task) volumeName(name string) string {
...
+		volumeMetadata := dockerClient.InspectVolume(ctx, vol.Name, dockerapi.InspectVolumeTimeout)
+		if volumeMetadata.Error != nil {
+			return errors.Wrap(volumeMetadata.Error, "initialize volume: auto provisioned volume detection failed")
+		}
+		return nil
+	}
+
+	// check if the volume configuration matches the one exists on the instance
+	volumeMetadata := dockerClient.InspectVolume(ctx, volumeConfig.DockerVolumeName, dockerapi.InspectVolumeTimeout)
+	if volumeMetadata.Error != nil {
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/task/task.go]: If the timeout error happened, that may mean the instance isn't in good condition to run tasks, the task would possibly fail. So I think we should prevent the task from running in this case.

File: agent/api/task/taskvolume_test.go [hunk 6]
```diff
@@ -19,52 +19,305 @@ import (
...
+	}
+
+	dockerClient.EXPECT().InspectVolume(gomock.Any(), gomock.Any(), gomock.Any()).Return(dockerapi.VolumeResponse{
+		DockerVolume: &docker.Volume{
+			Labels: map[string]string{"test": "test"},
+		},
+	})
+	err := testTask.initializeDockerVolumes(dockerClient, nil)
+	assert.Error(t, err, "volume resource details not match should cause task fail")
+}
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/api/task/taskvolume_test.go]: Added above.

File: agent/engine/engine_integ_test.go [hunk 7]
```diff
@@ -352,3 +353,43 @@ func TestEngineSynchronize(t *testing.T) {
...
+	testTask, tmpDirectory, err := createVolumeTask("shared", "TestSharedProvisionedVolume", "TestSharedProvisionedVolume", false)
+	defer os.Remove(tmpDirectory)
+	require.NoError(t, err, "creating test task failed")
+
+	go taskEngine.AddTask(testTask)
+
+	verifyTaskIsRunning(stateChangeEvents, testTask)
+	verifyTaskIsStopped(stateChangeEvents, testTask)
+	assert.Equal(t, *testTask.Containers[0].GetKnownExitCode(), 0)
+	assert.Equal(t, testTask.ResourcesMapUnsafe["dockerVolume"][0].(*taskresourcevolume.VolumeResource).VolumeConfig.DockerVolumeName, "TestSharedProvisionedVolume", "task volume name is not the same as specified in task definition")
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/engine/engine_integ_test.go]: added

**Pull request merged**